h2/src/main/org/h2/mvstore/tx/TransactionMap.java

```diff
@@ -871,109 +853,65 @@ void fetchNext() {
-
+            Snapshot snapshot = transactionMap.getStatementSnapshot();
+            uncommittedCursor = new Cursor<>(snapshot.root.root, from, to);
             fetchNext();
         }
 
         @Override
         void fetchNext() {
-            for (;;) {
+            current = null;
+            while (current == null) {
                 if (snapshotKey == null) {
                     fetchSnapshot();
                 }
                 if (uncommittedKey == null) {
                     fetchUncommitted();
                 }
-                if (snapshotKey != null) {
-                    if (uncommittedKey != null) {
-                        int cmp = keyType.compare(snapshotKey, uncommittedKey);
-                        if (cmp == 0) {
-                            if (uncommittedValue == null) {
-                                // This entry was deleted by this transaction
-                                snapshotKey = null;
-                                snapshotValue = null;
-                                uncommittedKey = null;
-                                continue;
-                            }
-                            // This entry was updated by this transaction, use the new value
-                            registerCurrent(uncommittedKey, uncommittedValue);
-                            snapshotKey = null;
-                            snapshotValue = null;
-                            uncommittedKey = null;
-                            uncommittedValue = null;
-                        } else if (cmp > 0) {
-                            if (uncommittedValue == null) {
-                                // Need to fetch the next uncommitted entry
-                                uncommittedKey = null;
-                                continue;
-                            }
-                            registerCurrent(uncommittedKey, uncommittedValue);
-                            uncommittedKey = null;
-                            uncommittedValue = null;
-                        } else {
-                            registerCurrent(snapshotKey, snapshotValue);
-                            snapshotKey = null;
-                            snapshotValue = null;
-                        }
-                    } else {
-                        registerCurrent(snapshotKey, snapshotValue);
-                        snapshotKey = null;
-                        snapshotValue = null;
-                    }
+                if (snapshotKey == null && uncommittedKey == null) {
+                    break;
+                }
+                int cmp = snapshotKey == null ? 1 :
+                            uncommittedKey == null ? -1 :
+                            keyType.compare(snapshotKey, uncommittedKey);
+                if (cmp < 0) {
+                    registerCurrent(snapshotKey, snapshotValue);
+                    snapshotKey = null;
                 } else {
```

Comment:
A `break` statement before else will be more efficient (no need to check `current` here), and it will allow to remove the `else { … }` around the remaining code and reduce its indentation.